src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionCache.java

```diff
@@ -183,25 +184,31 @@ private ListenableFuture<Void> uploadBlob(
       return remoteCacheClient.uploadBlob(context, digest, node.toByteString());
     }
 
-    PathOrBytes file = merkleTree.getFileByDigest(digest);
+    ContentSource file = merkleTree.getFileByDigest(digest);
     if (file != null) {
-      if (file.getBytes() != null) {
-        return remoteCacheClient.uploadBlob(context, digest, file.getBytes());
-      }
-
-      var path = checkNotNull(file.getPath());
-      try {
-        if (remotePathChecker.isRemote(context, path)) {
-          // If we get here, the remote input was determined to exist in the remote or disk cache at
-          // some point before action execution, but reported to be missing when querying the remote
-          // for missing action inputs; possibly because it was evicted in the interim.
-          reporter.post(new LostInputsEvent(digest));
-          throw new CacheNotFoundException(digest, path.getPathString());
+      return switch (file) {
+        case ContentSource.VirtualActionInputSource(VirtualActionInput virtualActionInput) ->
+            // TODO: Avoid materializing the entire file in memory. This requires changing the
+            // upload to be driven by an OutputStream rather than consuming an InputStream.
+            remoteCacheClient.uploadBlob(
+                context, digest, () -> virtualActionInput.getBytes().newInput());
```

Comment:
Might `virtualActionInput.getBytes()` be called multiple times by `Chunker`? The approach we use for the Google-internal RE is:

* Materialize the `ByteString` as late as possible, using double-checked locking, storing it as a field in the "supplier"
* The "supplier" has a `close()` method that is called when all chunks have been sent to the CAS. Calling this nulls the `ByteString` field.